src/fetchers.ts

```diff
@@ -298,28 +297,60 @@ async function fetchHelm(
 }
 
 async function fetchNixpkgs(
-  spec: NixpkgsFetch,
+  _spec: NixpkgsFetch,
   context: FetchContext,
 ): Promise<Version[]> {
-  const version = await fetchUrl("https://channels.nixos.org/" + spec.channel, context, undefined, { redirect: "manual" }, async (response) => {
-    const regexp = /releases.nixos.org\/nixpkgs\/([^/]+)/;
-    if (response.status !== 301) {
-      throw new Error(
-        `Got status ${response.status}: ${await response.text()}`,
-      );
-    }
-    const location = response.headers.get("Location");
-    if (location === null) {
-      throw new Error(`Could not get location for ${response.url}`);
-    }
-    const match = regexp.exec(location);
-    if (match === null) {
-      throw new Error(`Could not parse ${response.url}`);
+  const cacheKey = "nix-channel:nixpkgs-unstable";
+  let versions: string[] | undefined = undefined;
+  if (!context.update) {
+    const cached = context.cache.getBody(cacheKey);
+    if (cached) {
+      versions = JSON.parse(cached);
     }
-    return match[1];
-  });
+  }
+
+  if (versions === undefined) {
+    await dom.initParser();
+    versions = [];
+
+    let continuationToken: string | null = null;
+    do {
+      const url = new URL("https://nix-releases.s3.amazonaws.com");
+      url.searchParams.set("list-type", "2");
+      url.searchParams.set("prefix", "nixpkgs/nixpkgs-");
+      url.searchParams.set("delimiter", "/");
+      if (continuationToken !== null) {
+        url.searchParams.set("continuation-token", continuationToken);
+      }
```

Comment:
This roughly mirrors what http://releases.nixos.org/?prefix=nixpkgs/ does which is query the S3 API to get the listing of available versions. The prefix here restricts it to just `nixpkgs-unstable` essentially. Other channels like `nixos-22.11` use a different structure like `nixos/22.11/nixos-`.

Reply:
API documentation for this is here btw https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_RequestSyntax